File: Java/tests/hm/TestListComparator.java
package hm;

import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static org.assertj.core.api.Assertions.*;
import static org.junit.jupiter.api.Assertions.assertThrows;

class TestListComparator {
    private ListComparator comparator;

    /**
<<<<<<< HEAD
     * Test case for comparing empty lists and expecting an ArithmeticException with a specific message.
=======
     * Test case for comparing empty lists and
     * expecting an ArithmeticException with a specific message.
>>>>>>> dc5a0198b2dd8aa7c46f1431e8d4ab98893b3f6e
     */
    @Test
    void testNegativeEmptyList() {
        List<Integer> empty1 = new ArrayList<>();
        List<Integer> empty2 = new ArrayList<>();
        List<Integer> notEmpty = Arrays.asList(1, 2, 3);
        String expected = ListComparator.ZERO_EXCEPTION_MESSAGE;

        comparator = new ListComparator(empty1, empty2);
        Throwable throwable = assertThrows(ArithmeticException.class, () -> comparator.compareAverages());
        assertThat(throwable.getMessage()).isEqualTo(expected);

        comparator = new ListComparator(notEmpty, empty2);
        throwable = assertThrows(ArithmeticException.class, () -> comparator.compareAverages());
        assertThat(throwable.getMessage()).isEqualTo(expected);

        comparator = new ListComparator(empty1, notEmpty);
        throwable = assertThrows(ArithmeticException.class, () -> comparator.compareAverages());
        assertThat(throwable.getMessage()).isEqualTo(expected);
    }

    /**
<<<<<<< HEAD
     * Test case for comparing nullable list and
     * expecting an NullPointerException with a specific message.
     */
    @Test
    void testNegativeNullList() {
        List<Integer> notNullList = Arrays.asList(1, 2, 3);
        String expected = ListComparator.NULL_REFERENCE_EXCEPTION_MESSAGE;

        comparator = new ListComparator(null, null);
        Throwable throwable = assertThrows(NullPointerException.class, () -> comparator.compareAverages());
        assertThat(throwable.getMessage()).isEqualTo(expected);

        comparator = new ListComparator(null, notNullList);
        throwable = assertThrows(NullPointerException.class, () -> comparator.compareAverages());
        assertThat(throwable.getMessage()).isEqualTo(expected);

        comparator = new ListComparator(notNullList, null);
        throwable = assertThrows(NullPointerException.class, () -> comparator.compareAverages());
        assertThat(throwable.getMessage()).isEqualTo(expected);
    }

    /**
     * Test case for comparing lists where the average of the first list is higher than the second list.
=======
     * Test case for comparing lists
     * where the average of the first list is higher than the second list.
>>>>>>> dc5a0198b2dd8aa7c46f1431e8d4ab98893b3f6e
     */
    @Test
    void testPositiveFirstListAverageHigherThanSecond() {
        List<Integer> less = Arrays.asList(1, 2, 3);
        List<Integer> high = Arrays.asList(4, 9, 2);
        String expected = ListComparator.FIRST_LIST_AVG_HIGHER;
        comparator = new ListComparator(high, less);

        String actual = comparator.compareAverages();

        assertThat(actual).isEqualTo(expected);
    }

    /**
     * Test case for comparing lists
     * where the average of the second list is higher than the first list.
     */
    @Test
    void testPositiveFirstListAverageLessThanSecond() {
        List<Integer> less = Arrays.asList(1, 2, 3);
        List<Integer> high = Arrays.asList(4, 9, 2);
        String expected = ListComparator.SECOND_LIST_AVG_HIGHER;
        comparator = new ListComparator(less, high);

        String actual = comparator.compareAverages();

        assertThat(actual).isEqualTo(expected);
    }

    /**
     * Test case for comparing lists with equal averages.
     */
    @Test
    void testPositiveEqualListAverages() {
        List<Integer> first = Arrays.asList(1, 2, 3);
        List<Integer> second = Arrays.asList(1, 2, 3);
        String expected = ListComparator.LIST_ARE_EQUAL;
        comparator = new ListComparator(first, second);

        String actual = comparator.compareAverages();

        assertThat(actual).isEqualTo(expected);
    }
}

File: CSharp/tests/TestListComparator.cs
using NUnit.Framework;

namespace ListComaparatorWithTests;

public class TestListComparator
{

    /// <summary>
    /// Test case for comparing empty lists and expecting an ArithmeticException with a specific message.
    /// </summary>
    [Test]
    public void TestNegativeEmptyList()
    {
        List<int> empty = [];
        List<int> notEmpty = [1, 2, 3];

        string expected = ListComparator.ZERO_EXCEPTION_MESSAGE;

        ListComparator comparator = new(empty, empty);
        string actual = Assert.Throws<ArgumentException>(() => comparator.CompareAverages()).Message;
        Assert.That(actual, Is.EqualTo(expected));

        comparator = new(empty, notEmpty);
        actual = Assert.Throws<ArgumentException>(() => comparator.CompareAverages()).Message;
        Assert.That(actual, Is.EqualTo(expected));

        comparator = new(notEmpty, empty);
        actual = Assert.Throws<ArgumentException>(() => comparator.CompareAverages()).Message;
        Assert.That(actual, Is.EqualTo(expected));
    }

    /// <summary>
    ///  Test case for comparing nullable list and and expecting an NullReferenceException with a specific message.
    /// </summary>
    [Test]
    public void TestNegativeNullList()
    {
        List<int> notNull = [1, 2, 3];

        string expected = ListComparator.NULL_REFERENCE_EXCEPTION_MESSAGE;

        ListComparator comparator = new(null, null);
        string actual = Assert.Throws<NullReferenceException>(() => comparator.CompareAverages()).Message;
        Assert.That(actual, Is.EqualTo(expected));

        comparator = new(null, notNull);
        actual = Assert.Throws<NullReferenceException>(() => comparator.CompareAverages()).Message;
        Assert.That(actual, Is.EqualTo(expected));

        comparator = new(notNull, null);
        actual = Assert.Throws<NullReferenceException>(() => comparator.CompareAverages()).Message;
        Assert.That(actual, Is.EqualTo(expected));
    }

    /// <summary>
    /// Test case for comparing lists where the average of the first list is higher than the second list.
    /// </summary>
    [Test]
    public void TestPositiveFirstListAverageHigherThanSecond()
    {
        List<int> high = [4, 5, 6];
        List<int> less = [1, 2, 3];

        string expected = ListComparator.FIRST_LIST_AVG_HIGHER;

        ListComparator comparator = new(high, less);

        string actual = comparator.CompareAverages();
        Assert.That(actual, Is.EqualTo(expected));
    }

    /// <summary>
    /// Test case for comparing lists here the average of the second list is higher than the first list.
    /// </summary>
    [Test]
    public void TestPositiveFirstListAverageLessThanSecond()
    {
        List<int> high = [4, 5, 6];
        List<int> less = [1, 2, 3];

        string expected = ListComparator.SECOND_LIST_AVG_HIGHER;

        ListComparator comparator = new(less, high);

        string actual = comparator.CompareAverages();
        Assert.That(actual, Is.EqualTo(expected));

    }


    /// <summary>
    /// Test case for comparing lists with equal averages.
    /// </summary>
    [Test]
    public void TestPositiveEqualListAverages()
    {
        List<int> first = [1, 2, 3];
        List<int> second = [1, 2, 3];

        string expected = ListComparator.LIST_ARE_EQUAL;

        ListComparator comparator = new(first, second);

        string actual = comparator.CompareAverages();
        Assert.That(actual, Is.EqualTo(expected));
    }
}